staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/internal/modes/custom.go

```diff
@@ -0,0 +1,342 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package modes
+
+import (
+	"encoding"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
+
+	"github.com/fxamacker/cbor/v2"
+)
+
+// Returns a non-nil error if and only if the argument's type (or one of its component types, for
+// composite types) implements json.Marshaler or encoding.TextMarshaler without also implementing
+// cbor.Marshaler and likewise for the respective Unmarshaler interfaces. This will be removed in
+// favor of automatic transcoding between CBOR and JSON/text for these types.
+func RejectCustomMarshalers(v interface{}) error {
+	if v == nil {
+		return nil
+	}
+	rv := reflect.ValueOf(v)
+	if err := marshalerCache.getChecker(rv.Type()).check(rv, 128); err != nil {
+		return fmt.Errorf("unable to serialize %T: %w", v, err)
+	}
+	if err := unmarshalerCache.getChecker(rv.Type()).check(rv, 128); err != nil {
+		return fmt.Errorf("unable to serialize %T: %w", v, err)
+	}
+	return nil
+}
+
+// Recursion depth is limited as a basic mitigation against cyclic objects. Objects created by the
+// decoder shouldn't be able to contain cycles, but practically any object can be passed to the
+// encoder.
+var errMaxDepthExceeded = errors.New("object depth exceeds reasonable limits (possible cycle?)")
+
+var marshalerCache = checkers{
+	safeInterface: reflect.TypeFor[cbor.Marshaler](),
+	unsafeInterfaces: []reflect.Type{
+		reflect.TypeFor[json.Marshaler](),
+		reflect.TypeFor[encoding.TextMarshaler](),
+	},
+}
+
+var unmarshalerCache = checkers{
+	safeInterface: reflect.TypeFor[cbor.Unmarshaler](),
+	unsafeInterfaces: []reflect.Type{
+		reflect.TypeFor[json.Unmarshaler](),
+		reflect.TypeFor[encoding.TextUnmarshaler](),
+	},
+	assumeAddressableValues: true,
+}
+
+type checker struct {
+	check func(rv reflect.Value, depth int) error
+}
+
+// Having a single addressable checker for comparisons lets us prune and collapse parts of the
+// object traversal that are statically known to be safe.
+var noop = checker{
+	check: func(rv reflect.Value, depth int) error {
+		return nil
+	},
+}
+
+type checkers struct {
+	lock             sync.RWMutex
+	m                map[reflect.Type]**checker
+	safeInterface    reflect.Type
+	unsafeInterfaces []reflect.Type
+
+	assumeAddressableValues bool
+}
+
+func (cache *checkers) getChecker(rt reflect.Type) checker {
+	cache.lock.RLock()
+	c, ok := cache.m[rt]
+	cache.lock.RUnlock()
+	if ok {
+		return **c
+	}
```

Comment:
This is read heavy cache where each key is only written once?  Would `sync.Map` be more optimal here to minimize the read lock contention? (My understanding is that this is very high traffic, so it's not premature to hyper-optimize this)

Reply:
Yes, it looks like that would be a good fit here.